### PR TITLE
immer chat 4

### PR DESCRIPTION
**Pull request by chrisnojima**
- [x] unsent text
- [x] unfurl url map

### CODE REVIEW

File: shared/chat/conversation/messages/wrapper/unfurl/prompt-list/container.tsx [hunk 1]
```diff
@@ -11,74 +10,56 @@ type OwnProps = {
...
+      promptDomains = pm.get(messageID)
+    }
+    return {
+      messageID,
+      promptDomains: promptDomains || noPrompts,
+    }
   },
-})
-
-const makeRes = (actionType: RPCChatTypes.UnfurlPromptAction, domain?: string) => {
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/wrapper/unfurl/prompt-list/container.tsx]: this wasn't typed correctly. the `onetime`/`accept` fields are only acceptable on those `actionType`s

File: shared/reducers/chat2.tsx [hunk 2]
```diff
@@ -499,11 +499,18 @@ export default (_state: Types.State = initialState, action: Actions): Types.Stat
...
+        const mmap = new Map(unfurlPromptMap.get(conversationIDKey) || [])
+        const prompts = new Set(mmap.get(messageID) || [])
+        if (show) {
+          prompts.add(domain)
+        } else {
+          prompts.delete(domain)
+        }
+        mmap.set(messageID, prompts)
+        unfurlPromptMap.set(conversationIDKey, mmap)
+        draftState.unfurlPromptMap = unfurlPromptMap
```
Comment (buoyad) [inline shared/reducers/chat2.tsx]: These are tricky. I wonder if some helper like `updateMap` could help. Maybe like
```
draftState.unfurlPromptMap = updateMap(draftState.unfurlPromptMap, (map) => {
  const mmap = updateMap(unfurlPromptMap.get(conversationIDKey) || [], (mmap) =>  {
    const prompts = new Set(mmap.get(messageID) || [])
    if (show) {
      prompts.add(domain)
    } else {
      prompts.delete(domain)
    }
    return mmap.set(messageID, prompts)
  })
  return map.set(conversationIDKey, mmap)
})
```
Where `updateMap` calls `new Map` on the first param. It's not much simpler code-wise but I think mentally it's easier to reach for `updateMap` every time you change a map in a reducer. Anyway, just an idea.

Comment (chrisnojima) [reply shared/reducers/chat2.tsx]: i think its hard to get this to type correctly. hopefully a bunch of this goes away with 
`?.` and Map support in the next immer

Comment (chrisnojima) [reply shared/reducers/chat2.tsx]: its also likely simpler to just not have all these nested things at all and just flatten it out. just make the keys `convoid + : + messageid`

Comment (buoyad) [reply shared/reducers/chat2.tsx]: Yeah agreed we should try to avoid nesting these

**Pull request merged**